### PR TITLE
inspection: use pep517 metadata build

### PR DESCRIPTION
**Pull request by abn**
This change replaces `setup.py` execution in favour of pep517 metadata builds. The proposed approach executes `pep517.meta.build` instead of `python setup.py egg_info`.

This does not change the order of preference in the inspection flow for execution vs AST parsing. A build is only perfromed in cases where existing metadata inspection, `SetupReader` parsing etc returns an incomplete/unusable `PackageInfo` instance or raises `PackageInfoError`.

The original proposal was made in #2296 by @finswimmer.

Closes: #2296 
Resolves: #1900

# Pull Request Check List

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

### CODE REVIEW

File: poetry/inspection/info.py [hunk 1]
```diff
@@ -392,10 +401,44 @@ def _get_poetry_package(path):  # type: (Path) -> Optional[ProjectPackage]
...
+        try:
+            info = cls.from_setup_files(path)
+            if info.requires_dist is not None:
+                return info
+        except PackageInfoError:
+            pass
+
+        try:
+            with temporary_directory() as tmp_dir:
+                pep517_metadata_build(
```
Comment (sdispater) [inline poetry/inspection/info.py:421]: The issue I have with this is that `pep517` uses `sys.executable` (i.e. the Python interpreter used to run Poetry) and, thus, is not compatible with Poetry's environment system.

That's why I mentioned in https://github.com/python-poetry/poetry/pull/2296#issuecomment-636839053 that we would most likely not be able to use `pep517`.

Comment (abn) [reply poetry/inspection/info.py:421]: We can workaround that by executing `python -m pep517.meta.build` under the project's environment.

Comment (sdispater) [reply poetry/inspection/info.py:421]: The python version of Poetry's environment is not necessarily the same as the one from the project's environment so if a build backend relies on it, and this is particularly true for setuptools, we might end up with the wrong information.

Comment (abn) [reply poetry/inspection/info.py]: @sdispater I have added a commit wherein we retail the old behaviour wrt venv before building metadata.

**Comment by abn** [comment]
@sdispater bumping this once again with the proposed changes addressing your concerns

**Pull request merged**